dm/dm/worker/server.go

```diff
@@ -299,43 +303,49 @@ func (s *Server) observeRelayConfig(ctx context.Context, rev int64) error {
 						break
 					}
 					rev = rev1
-					if relaySource == nil {
-						if w := s.getSourceWorker(true); w != nil && w.startedRelayBySourceCfg {
-							break
+
+					// compare relaySources and source_workers
+					// 1. no relaySource but has source_worker, try to disableRelay
+					// 2. has relaySource but no source_worker, try to starworker
+					// 3. has relaySource and has source_worker, check and try to enableRelay
+					ws := s.getSourceWorkers(true)
+					consistents := make([]string, 0)
+					for _, relaySource := range relaySources {
+						// has relaySource but no worker
+						if w, ok := ws[relaySource.SourceID]; !ok {
+							log.L().Info(fmt.Sprintf("didn't find source_worker but found relay config %s after etcd retryable error. Will start source_worker", relaySource.SourceID))
+							if err2 := s.enableRelay(relaySource, true); err2 != nil {
+								return err2
+							}
+						} else {
+							// has relaySource and worker
+							log.L().Info(fmt.Sprintf("found source_worker and relay config %s after etcd retryable error. Will check and enable relay", relaySource.SourceID))
+							consistents = append(consistents, relaySource.SourceID)
+							if !w.relayEnabled.Load() {
+								if err2 := w.EnableRelay(false); err2 != nil {
+									return err2
+								}
+							}
 						}
-						log.L().Info("didn't found relay config after etcd retryable error. Will stop relay now")
-						err = s.disableRelay("")
-						if err != nil {
-							log.L().Error("fail to disableRelay after etcd retryable error", zap.Error(err))
-							return err // return if failed to stop the worker.
+					}
+					// has worker but no relaySource
+					for _, w := range ws {
+						needDisableRelay := true
+						for _, consistent := range consistents {
+							if w.cfg.SourceID == consistent {
+								needDisableRelay = false
+								break
+							}
 						}
```

Comment:
To me, we don't need a `consistents` array? Just stop the worker's who is not in `relaySources` and `w.startedRelayBySourceCfg==false` is enough. To do this, we may save `relaySources` as a map in `ha` package

Reply:
yes, `consistents` is a array and works just like `relaySources`，it will be better for traversal if `relaySources` is a map. i will fix it.